src/cargo/ops/cargo_new.rs

```diff
@@ -44,40 +56,181 @@ struct CargoNewConfig {
-                        new_name);
-                    try!(config.shell().say(&message, BLACK));
-                }
-                new_name
-            }
+    
+    let dir_name = try!(path.file_name().and_then(|s| s.to_str()).chain_error(|| {
+        human(&format!("cannot create a project with a non-unicode name: {:?}",
+                       path.file_name().unwrap()))
+    }));
+    
+    if opts.bin {
+        Ok(dir_name)
+    } else {
+        let new_name = strip_rust_affixes(dir_name);
+        if new_name != dir_name {
+            let message = format!(
+                "note: package will be named `{}`; use --name to override",
+                new_name);
+            try!(config.shell().say(&message, BLACK));
         }
-    };
+        Ok(new_name)
+    }
+}
+
+fn check_name(name: &str) -> CargoResult<()> {
     for c in name.chars() {
         if c.is_alphanumeric() { continue }
         if c == '_' || c == '-' { continue }
         return Err(human(&format!("Invalid character `{}` in crate name: `{}`",
                                   c, name)));
     }
-    mk(config, &path, name, &opts).chain_error(|| {
+    Ok(())
+}
+
+fn detect_source_paths_and_types<'a : 'b, 'b>(project_path : &'a Path, 
+                                            project_name: &'a str, 
+                                            detected_files: &'b mut Vec<SourceFileInformation>) 
+                                                -> CargoResult<()> {
+    let path = project_path;
+    let name = project_name;
+    
+    enum H {
+        Bin,
+        Lib,
+        Detect,
+    }
+    
+    struct Test {
+        proposed_path: String,
+        handling: H,
+    }
+        
+    let tests = vec![
+        Test { proposed_path: format!("src/main.rs"),     handling: H::Bin },
+        Test { proposed_path: format!("main.rs"),         handling: H::Bin },
+        Test { proposed_path: format!("src/{}.rs", name), handling: H::Detect },
+        Test { proposed_path: format!("{}.rs", name),     handling: H::Detect },
+        Test { proposed_path: format!("src/lib.rs"),      handling: H::Lib },
+        Test { proposed_path: format!("lib.rs"),          handling: H::Lib },
+    ];
+    
+    for i in tests {
+        let pp = i.proposed_path;
+        if ! paths::file_already_exists(&path.join(pp.clone())) {
+            continue;
+        }
+        let sfi = match i.handling {
+            H::Bin => 
+                SourceFileInformation { relative_path: pp, bin: true },
+            H::Lib => 
+                SourceFileInformation { relative_path: pp, bin: false },
+            H::Detect => {
+                let content = try!(paths::read(&path.join(pp.clone())));
+                let isbin = content.contains("fn main");
+                SourceFileInformation { relative_path: pp, bin: isbin }
+            },
+        };
+        detected_files.push(sfi);
+    }
+    
+    Ok(())
+}
+
+fn plan_new_source_file(bin: bool) -> SourceFileInformation {
```

Comment:
No, it something around "add creation of new source file to the queue" or "get default SourceFileInformation depending on binary-ness" or "schedule creation of a new source file".